crates/starknet-types-core/src/short_string/mod.rs

```diff
@@ -0,0 +1,179 @@
+//! A `ShortString` is string that have been checked and is guaranted to be convertible into a valid `Felt`.
+//! It checks that the `String` only contains ascii characters and is no longer than 31 characters.
+//!
+//! The convesion to `Felt` is done by using the internal ascii short string as bytes and parse those as a big endian number.
+
+#[cfg(not(feature = "std"))]
+use crate::felt::alloc::string::{String, ToString};
+use crate::felt::Felt;
+
+/// A cairo short string
+///
+/// Allow for safe conversion of cairo short string `String` into `Felt`,
+/// as it is guaranted that the value it contains can be represented as a felt.
+#[repr(transparent)]
+#[derive(Debug, Clone, Default, PartialEq, Eq, PartialOrd, Ord)]
+pub struct ShortString(String);
+
+impl core::fmt::Display for ShortString {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl From<ShortString> for Felt {
+    fn from(ss: ShortString) -> Self {
+        let bytes = ss.0.as_bytes();
+
+        let mut buffer = [0u8; 32];
+        buffer[(32 - bytes.len())..].copy_from_slice(bytes);
+
+        // The conversion will never fail
+        Felt::from_bytes_be(&buffer)
+    }
+}
+
+#[derive(Debug, Clone)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+pub enum TryShortStringFromStringError {
+    TooLong,
+    NonAscii,
+}
+
+impl core::fmt::Display for TryShortStringFromStringError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            TryShortStringFromStringError::TooLong => "string to long",
+            TryShortStringFromStringError::NonAscii => "string contains non ascii characters",
+        }
+        .fmt(f)
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for TryShortStringFromStringError {}
+
+impl TryFrom<String> for ShortString {
+    type Error = TryShortStringFromStringError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        if value.len() > 31 {
+            return Err(TryShortStringFromStringError::TooLong);
+        }
+        if !value.as_str().is_ascii() {
+            return Err(TryShortStringFromStringError::NonAscii);
+        }
+
+        Ok(ShortString(value))
+    }
+}
+
+impl Felt {
+    /// Create a felt value from a cairo short string
+    ///
+    /// The string must contains only ascii characters
+    /// and its length must not exceed 31.
+    pub fn try_from_cairo_short_string(
+        string: &str,
+    ) -> Result<Self, TryShortStringFromStringError> {
+        if string.len() > 31 {
+            return Err(TryShortStringFromStringError::TooLong);
+        }
+        let bytes = string.as_bytes();
+        if !bytes.is_ascii() {
+            return Err(TryShortStringFromStringError::NonAscii);
+        }
+
+        let mut buffer = [0u8; 32];
+        buffer[(32 - bytes.len())..].copy_from_slice(bytes);
+
+        // The conversion will never fail
+        Ok(Felt::from_bytes_be(&buffer))
+    }
```

Comment:
why not using the `TryFrom` impl ? and naming it `try_from_cairo_short_string` is confusing imo as you're expecting a `&str`

Reply:
`try_from_cairo_short_string` you find it confusing because it takes a `&str` but the name says `string`.

We are not referring to rust `String` type here but to "cairo short string?, which is another concept, that just happen to also use the the word "string". Not my choice of naming, but that is what those are called.
Mentioning that those are "cairo short" help for the disambiguation tho

Reply:
it's confusing because of what gabriel explained. It's called from cairo short string but doesn't expect a short string. Imo this should be `impl From<ShortString> for Felt`

Reply:
@0xLucqs `impl From<ShortString> for Felt` do already exists. But that means that you will have to manually do:
```rust
let sn_sepolia_id = Felt::from(ShortString::try_from("SN_SEPOLIA")?);
```
which is verbose and requires understanding the impl detail that `ShortString` is.

This function is a handy shortcut for developers that want to parse a cairo short string.
```rust
let sn_sepolia_id = Felt::try_from_cairo_short_string("SN_SEPOLIA")?;
```

It's expressive, it's straightforward, and it is even a bit more optimized as it avoids the allocation of a new string, which comes with the initialization of a `ShortString`

Reply:
@gabrielbosio 

> One should expect to pass a ShortString as an argument to this function, not a &str.

Yeah I get your point.

> /// Create a felt value from a string. This function converts the passed string to a `ShortString` and then it converts the result to a `Felt`, returning it.

The only issue is that it is not what the implementation does. It doesn't internally create a `ShortString` type, because it doesn't need to.

Reply:
`short_string!("some short string");`

Reply:
This could be a great addition, to execute at compile time and crash if the provided value is not valid.
But we also should provide a solution for runtime parsing of `String` (`parse_cairo_short_string(&a_string)`)

Reply:
No, we don't want to use the intermediate `ShortString` type there. It is suboptimal because it requires creating `String`, which will involve allocation if input is a `&str`.
Those are two slightly different use cases.

Also, it is way less user-friendly and expressive than `Felt::parse_cairo_short_string(&a_string)` which describes exactly what it does